cmake/kokkos_functions.cmake

```diff
@@ -60,6 +60,235 @@ FUNCTION(kokkos_option CAMEL_SUFFIX DEFAULT TYPE DOCSTRING)
+  
+  GET_TARGET_PROPERTY(TPL_COMPILE_OPTIONS ${NAME} INTERFACE_COMPILE_OPTIONS)
+  SET(TPL_LINK_OPTIONS)
+  IF(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.13.0")
+    GET_TARGET_PROPERTY(TPL_LINK_OPTIONS ${NAME} INTERFACE_LINK_OPTIONS)
+  ENDIF()
+  GET_TARGET_PROPERTY(TPL_LINK_LIBRARIES  ${NAME} INTERFACE_LINK_LIBRARIES)
+
+  GET_TARGET_PROPERTY(LIB_TYPE ${NAME} TYPE)
+  IF (NOT ${LIB_TYPE} STREQUAL "INTERFACE_LIBRARY")
+    GET_TARGET_PROPERTY(TPL_LIBRARY ${NAME} IMPORTED_LOCATION)
+  ENDIF()
+  GET_TARGET_PROPERTY(TPL_INCLUDES ${NAME} INTERFACE_INCLUDE_DIRECTORIES)
+
+  IF(TPL_LIBRARIES)
+    KOKKOS_APPEND_CONFIG_LINE("IMPORTED_LOCATION ${TPL_LIBRARIES}")
+  ENDIF()
+  IF(TPL_INCLUDES)
+    KOKKOS_APPEND_CONFIG_LINE("INTERFACE_INCLUDE_DIRECTORIES ${TPL_INCLUDES}")
+  ENDIF()
+  IF(TPL_COMPILE_OPTIONS)
+    KOKKOS_APPEND_CONFIG_LINE("INTERFACE_COMPILE_OPTIONS ${TPL_COMPILE_OPTIONS}")
+  ENDIF()
+  IF(TPL_LINK_OPTIONS)
+    KOKKOS_APPEND_CONFIG_LINE("INTERFACE_LINK_OPTIONS ${TPL_LINK_OPTIONS}")
+  ENDIF()
+  IF(TPL_LINK_LIBRARIES)
+    KOKKOS_APPEND_CONFIG_LINE("INTERFACE_LINK_LIBRARIES ${TPL_LINK_LIBRARIES}")
+  ENDIF()
+  KOKKOS_APPEND_CONFIG_LINE(")")
+ENDMACRO()
+
+
+MACRO(kokkos_import_tpl MODULE_NAME)
+  CMAKE_PARSE_ARGUMENTS(TPL
+   "NO_EXPORT"
+   "OPTION_NAME;IMPORTED_NAME"
+   ""
+   ${ARGN})
+  IF (NOT TPL_OPTION_NAME)
+    SET(TPL_OPTION_NAME ${MODULE_NAME})
+  ENDIF()
+  IF (NOT TPL_IMPORTED_NAME)
+    SET(TPL_IMPORTED_NAME Kokkos::${TPL_OPTION_NAME})
+  ENDIF()
+  IF(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.12.0") 
+    CMAKE_POLICY(SET CMP0074 NEW)
+  ENDIF()
+  IF (KOKKOS_ENABLE_${TPL_OPTION_NAME})
+    FIND_PACKAGE(${MODULE_NAME} REQUIRED MODULE)
+    IF(NOT TARGET ${TPL_IMPORTED_NAME})
+      MESSAGE(FATAL_ERROR "Find module succeeded for ${MODULE_NAME}, but did not produce valid target ${TPL_IMPORTED_NAME}")
+    ENDIF()
+    IF(NOT TPL_NO_EXPORT)
+      KOKKOS_EXPORT_IMPORTED_TPL(${TPL_IMPORTED_NAME})
+    ENDIF()
+  ENDIF()
+ENDMACRO(kokkos_import_tpl)
+
+MACRO(kokkos_import_cmake_tpl MODULE_NAME)
+  kokkos_import_tpl(${MODULE_NAME} ${ARGN} NO_EXPORT)
+  CMAKE_PARSE_ARGUMENTS(TPL
+   "NO_EXPORT"
+   "OPTION_NAME"
+   ""
+   ${ARGN})
+
+  IF (NOT TPL_OPTION_NAME)
+    SET(TPL_OPTION_NAME ${MODULE_NAME})
+  ENDIF()
+
+  IF (NOT TPL_NO_EXPORT)
+    KOKKOS_EXPORT_CMAKE_TPL(${MODULE_NAME})
+  ENDIF()
+ENDMACRO()
+
+MACRO(kokkos_create_imported_tpl NAME)
+  CMAKE_PARSE_ARGUMENTS(TPL
+   "INTERFACE"
+   "LIBRARY"
+   "LINK_LIBRARIES;INCLUDES;COMPILE_OPTIONS;LINK_OPTIONS"
+   ${ARGN})
+
+  IF (TPL_INTERFACE)
+    ADD_LIBRARY(${NAME} INTERFACE IMPORTED)
+  ELSE()
+    ADD_LIBRARY(${NAME} UNKNOWN IMPORTED)
+  ENDIF()
+
+  IF(TPL_LIBRARY)
+    SET_TARGET_PROPERTIES(${NAME} PROPERTIES
+      IMPORTED_LOCATION ${TPL_LIBRARY})
+  ENDIF()
+
+  #Things have to go in quoted in case we have multiple list entries
+  IF(TPL_LINK_LIBRARIES)
+    SET_TARGET_PROPERTIES(${NAME} PROPERTIES
+      INTERFACE_LINK_LIBRARIES "${TPL_LINK_LIBRARIES}")
```

Comment:
I am getting
```
CMake Error at cmake/kokkos_functions.cmake:175 (SET_TARGET_PROPERTIES):
  Property INTERFACE_LINK_LIBRARIES may not contain link-type keyword
  "general".  The INTERFACE_LINK_LIBRARIES property may contain
  configuration-sensitive generator-expressions which may be used to specify
  per-configuration rules.
Call Stack (most recent call first):
  cmake/kokkos_tpls.cmake:45 (KOKKOS_CREATE_IMPORTED_TPL)
  cmake/kokkos_tribits.cmake:186 (INCLUDE)
  CMakeLists.txt:134 (KOKKOS_SETUP_BUILD_ENVIRONMENT)
```
with `CMake 3.10.2`.

Reply:
I see this only with the (experimental) `HPX` backend.

Reply:
Was the HPX stuff ever working? I really didn't change anything. I'm still linking to all the same libraries contained in `HPX_LIBRARIES`. HPX does not export a library named `general`.

Reply:
Yes, I was testing it before. It's not necessarily a bug here, but it would still be good if we could get it working for the release.

Reply:
So the only thing that's different is that instead of directly calling
````
target_link_libraries(kokkoscore ${HPX_LIBRARIES})
````
I stash `HPX_LIBRARIES` in `Kokkos::HPX` in `INTERFACE_LINK_LIBRARIES`.
If `general` is not a valid CMake target, then that would mean CMake would append a random `-lgeneral` to the link line.  Is there a `libgeneral` on the system? I don't understand how this could have been working.

Reply:
I opened an issue with HPX.

https://github.com/STEllAR-GROUP/hpx/issues/4144

Reply:
An example link line I am seeing (without this pull request) is
> [...]
> -Wl,-rpath,/home/darndt/hpx-install/lib:/home/darndt/boost-install-69/lib:/home/darndt/hwloc-install/lib libkokkos_gtest.a ../../containers/src/libkokkoscontainers.a ../src/libkokkoscore.a /home/darndt/hpx-install/lib/libhpx_init.a /home/darndt/hpx-install/lib/libhpx.so.1.3.0 -latomic -ldl -lrt /home/darndt/boost-install-69/lib/libboost_filesystem.so /home/darndt/boost-install-69/lib/libboost_program_options.so /home/darndt/boost-install-69/lib/libboost_system.so -lpthread /home/darndt/boost-install-69/lib/libboost_filesystem.so /home/darndt/boost-install-69/lib/libboost_program_options.so /home/darndt/boost-install-69/lib/libboost_system.so -lpthread /home/darndt/hwloc-install/lib/libhwloc.so /usr/lib/x86_64-linux-gnu/libdl.so
```

Reply:
I'm officially confused.

Reply:
I cannot fathom how the list of `HPX_LIBRARIES` exported leads to that link line.

Reply:
Nevermind. I'm an idiot that shouldn't be allowed around CMake.

Reply:
Stripping general by something like
```CMake
diff --git a/cmake/kokkos_tpls.cmake b/cmake/kokkos_tpls.cmake
index 59b03538..5ceb7d3d 100644
--- a/cmake/kokkos_tpls.cmake
+++ b/cmake/kokkos_tpls.cmake
@@ -42,6 +42,9 @@ IF (KOKKOS_ENABLE_HPX)
   FIND_PACKAGE(HPX REQUIRED)
   #as of right now, HPX doesn't export correctly
   #so let's convert it to an imported target
+
+  string(REGEX REPLACE "general" "" HPX_LIBRARIES "${HPX_LIBRARIES}")
+
   KOKKOS_CREATE_IMPORTED_TPL(Kokkos::HPX INTERFACE
     LINK_LIBRARIES ${HPX_LIBRARIES}
     INCLUDES ${HPX_INCLUDE_DIRS}
```
would make that work again.